File: frontend/src/TerraForge/lib/combine.js
// Combine two heightmaps with various blend modes.
export const BLEND_MODES = ["add", "multiply", "max", "min", "lerp", "screen", "subtract"];

export function combineHeightmaps(a, b, size, mode = "lerp", blend = 0.5) {
  if (a.length !== b.length) throw new Error("size mismatch");
  const out = new Float32Array(a.length);
  for (let i = 0; i < a.length; i++) {
    const va = a[i];
    const vb = b[i];
    let v;
    switch (mode) {
      case "add": v = va + vb * blend; break;
      case "subtract": v = va - vb * blend; break;
      case "multiply": v = va * (1 - blend + vb * blend); break;
      case "max": v = Math.max(va, vb * blend + va * (1 - blend)); break;
      case "min": v = Math.min(va, vb * blend + va * (1 - blend)); break;
      case "screen": v = 1 - (1 - va) * (1 - vb * blend); break;
      case "lerp":
      default: v = va * (1 - blend) + vb * blend; break;
    }
    out[i] = Math.max(0, Math.min(1, v));
  }
  return out;
}

File: TerraForgeRust/core/src/combine.rs
//! Heightmap blend modes. Port of `lib/combine.js`.

#[derive(Clone, Copy, Debug)]
pub enum BlendMode { Add, Subtract, Multiply, Max, Min, Screen, Lerp }

impl BlendMode {
    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "add" => Some(Self::Add),
            "subtract" => Some(Self::Subtract),
            "multiply" => Some(Self::Multiply),
            "max" => Some(Self::Max),
            "min" => Some(Self::Min),
            "screen" => Some(Self::Screen),
            "lerp" => Some(Self::Lerp),
            _ => None,
        }
    }
}

pub fn combine_heightmaps(a: &[f32], b: &[f32], mode: BlendMode, blend: f32) -> Vec<f32> {
    assert_eq!(a.len(), b.len());
    let mut out = vec![0f32; a.len()];
    for i in 0..a.len() {
        let va = a[i];
        let vb = b[i];
        let v = match mode {
            BlendMode::Add => va + vb * blend,
            BlendMode::Subtract => va - vb * blend,
            BlendMode::Multiply => va * (1.0 - blend + vb * blend),
            BlendMode::Max => va.max(vb * blend + va * (1.0 - blend)),
            BlendMode::Min => va.min(vb * blend + va * (1.0 - blend)),
            BlendMode::Screen => 1.0 - (1.0 - va) * (1.0 - vb * blend),
            BlendMode::Lerp => va * (1.0 - blend) + vb * blend,
        };
        out[i] = v.clamp(0.0, 1.0);
    }
    out
}
